File: src/audio_stegano_desktop/src/AudioStegano.Desktop/Services/PayloadImageCodec.cs
using System.IO;
using System.Windows.Media.Imaging;
using AudioStegano.Core.Stego;

namespace AudioStegano.Desktop.Services;

/// <summary>
/// Compresses a user-picked still image into a JPEG body that fits the ASTG bit budget.
/// Uses WPF imaging (no extra NuGet).
/// </summary>
public static class PayloadImageCodec
{
    public static byte[] CompressToFitBudget(byte[] sourceBytes, int bitBudget)
    {
        var maxBytes = PayloadImageDefaults.MaxImageBytesForBitBudget(bitBudget);
        if (maxBytes < 64)
            throw new InvalidOperationException("Bit budget too small for an image payload");

        BitmapSource bitmap;
        using (var ms = new MemoryStream(sourceBytes))
        {
            var decoder = BitmapDecoder.Create(
                ms,
                BitmapCreateOptions.PreservePixelFormat,
                BitmapCacheOption.OnLoad);
            bitmap = decoder.Frames[0];
        }

        var longEdge = Math.Max(bitmap.PixelWidth, bitmap.PixelHeight);
        if (longEdge > PayloadImageDefaults.MaxLongEdgePx)
        {
            var scale = (double)PayloadImageDefaults.MaxLongEdgePx / longEdge;
            bitmap = new TransformedBitmap(
                bitmap,
                new System.Windows.Media.ScaleTransform(scale, scale));
            bitmap.Freeze();
        }

        for (var edge = PayloadImageDefaults.MaxLongEdgePx;
             edge >= 64;
             edge = Math.Clamp((int)(edge * 0.75), 64, PayloadImageDefaults.MaxLongEdgePx))
        {
            var candidate = bitmap;
            var currentLong = Math.Max(candidate.PixelWidth, candidate.PixelHeight);
            if (currentLong > edge)
            {
                var scale = (double)edge / currentLong;
                candidate = new TransformedBitmap(
                    bitmap,
                    new System.Windows.Media.ScaleTransform(scale, scale));
                candidate.Freeze();
            }

            for (var q = PayloadImageDefaults.JpegQuality;
                 q >= PayloadImageDefaults.MinJpegQuality;
                 q -= 10)
            {
                var jpeg = EncodeJpeg(candidate, q);
                if (jpeg.Length <= maxBytes)
                    return jpeg;
            }
        }

        throw new InvalidOperationException(
            $"Image cannot be compressed under the bit budget ({maxBytes} bytes)");
    }

    static byte[] EncodeJpeg(BitmapSource source, int qualityPercent)
    {
        var encoder = new JpegBitmapEncoder
        {
            QualityLevel = Math.Clamp(qualityPercent, 1, 100),
        };
        encoder.Frames.Add(BitmapFrame.Create(source));
        using var ms = new MemoryStream();
        encoder.Save(ms);
        return ms.ToArray();
    }
}

File: src/audio_stegano_app/lib/core/stego/payload_image_codec.dart
import 'dart:typed_data';

import 'package:image/image.dart' as img;

import 'payload_envelope.dart';

/// Compresses a user-picked still image into a JPEG body that fits the
/// ASTG bit budget (long-edge shrink + quality ladder).
abstract final class PayloadImageCodec {
  /// Returns JPEG bytes sized for [bitBudget], or throws if even the smallest
  /// attempt exceeds the budget.
  static Uint8List compressToFitBudget(
    Uint8List sourceBytes, {
    required int bitBudget,
  }) {
    final maxBytes = PayloadImageDefaults.maxImageBytesForBitBudget(bitBudget);
    if (maxBytes < 64) {
      throw StateError('Bit budget too small for an image payload');
    }

    final decoded = img.decodeImage(sourceBytes);
    if (decoded == null) {
      throw const FormatException('Unsupported or corrupt image file');
    }

    var working = decoded;
    final longEdge = working.width > working.height ? working.width : working.height;
    if (longEdge > PayloadImageDefaults.maxLongEdgePx) {
      working = img.copyResize(
        working,
        width: working.width >= working.height
            ? PayloadImageDefaults.maxLongEdgePx
            : null,
        height: working.height > working.width
            ? PayloadImageDefaults.maxLongEdgePx
            : null,
        interpolation: img.Interpolation.average,
      );
    }

    for (var edge = PayloadImageDefaults.maxLongEdgePx;
        edge >= 64;
        edge = (edge * 0.75).round().clamp(64, PayloadImageDefaults.maxLongEdgePx)) {
      var candidate = working;
      final currentLong =
          candidate.width > candidate.height ? candidate.width : candidate.height;
      if (currentLong > edge) {
        candidate = img.copyResize(
          working,
          width: working.width >= working.height ? edge : null,
          height: working.height > working.width ? edge : null,
          interpolation: img.Interpolation.average,
        );
      }

      for (var q = PayloadImageDefaults.jpegQuality;
          q >= PayloadImageDefaults.minJpegQuality;
          q -= 10) {
        final jpeg = Uint8List.fromList(
          img.encodeJpg(candidate, quality: q),
        );
        if (jpeg.length <= maxBytes) {
          return jpeg;
        }
      }
    }

    throw StateError(
      'Image cannot be compressed under the bit budget ($maxBytes bytes)',
    );
  }
}
